lib/cloud_formation/executor.rb

```diff
@@ -52,25 +52,42 @@ def create_change_set(type: "UPDATE")
     end
 
     def template_name
-      "stack_templates/#{stack.name}/#{Time.current.strftime("%Y-%m-%d-%H%M%S")}.template"
+      @template_name ||= "stack_templates/#{stack.name}/#{Time.current.strftime("%Y-%m-%d-%H%M%S")}.template"
     end
 
     def upload_to_s3!
-      resp = @s3_client.put_object({
-        body: stack.target!,
+      params = {
         bucket: @bucket,
         key: template_name,
+      }
+
+      resp = @s3_client.put_object({
+        body: stack.target!,
+        **params
       })
-      Rails.logger.info "Uploaded stack template to bucket"
       Rails.logger.info resp
+
+      Rails.logger.info "Waiting for stack template to be uploaded"
+      begin
+        @s3_client.wait_until(:object_exists, params, 
+          before_wait: -> (attempts, response) do
+            Rails.logger.info "Waiting for stack template to be uploaded"
+          end
+        )
+      rescue Aws::Waiters::Errors::WaiterFailed => e
+        Rails.logger.warn "Upload failed: #{e.message}"
+        raise e
+      end
+
+      Rails.logger.info "Uploaded stack template to bucket"
     end
 
     def stack_options
       upload_to_s3!
       {
         stack_name: stack.name,
         capabilities: ["CAPABILITY_IAM"],
-        template_url: "https://#{@bucket}.s3.amazonaws.com/#{template_name}"
+        template_url: "https://s3.#{stack.region}.amazonaws.com/#{@bucket}/#{template_name}"
       }
     end
```

Comment:
@davidsiaw 
How about making `template_name` local variable instead of an instance_method?
```ruby
def stack_options
  template_name = "stack_templates/#{stack.name}/#{Time.current.strftime("%Y-%m-%d-%H%M%S")}.template"
  upload_to_s3!(template_name)
  {
        stack_name: stack.name,
        capabilities: ["CAPABILITY_IAM"],
        template_url: "https://s3.#{stack.region}.amazonaws.com/#{@bucket}/#{template_name}"
  }
end
```

I think this makes more clear about the role or scope of the `template_name`.

- It will be used only for communication between uploading and executing
- We don't have to expose it to other classes or save it for future use

Reply:
Its used in two places. if I do that the code will be duplicated.

Reply:
@essa Its used in two places. if I do that the code will be duplicated.

Reply:
@davidsiaw 
I meant removing the method and making it a local variable for passing as a parameter.
Where is the duplication?

Reply:
@essa my bad, I misread your code. That is a good idea. will do that.